Define instance vst2 2nd

pietro68 opened the pull request:
Resolves: https://github.com/audacity/audacity/issues/3192

<!-- Use "x" to fill the checkboxes below like [x] -->

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -101,6 +101,10 @@ struct VSTEffectSettings
 
 struct VSTEffectWrapper : public VSTEffectLink, public XMLTagHandler
 {
+   VSTEffectWrapper(const PluginPath& path)
```
Most one-argument constructors should be `explicit`

pietro68 replied:
ok, done

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -218,6 +222,34 @@ struct VSTEffectWrapper : public VSTEffectLink, public XMLTagHandler
+   bool     mAutomatable;
+
+   virtual void callSetProgram(int index) = 0;
+   virtual void Unload() = 0;
```
I see you add two virtual functions to the wrapper when you moved Load() into it.

I understand virtual `Unload` lets wrapper avoid mention of `mDialog`.

But I think the last part of `Unload`:

```
   if (mModule)
   {
#if defined(__WXMAC__)
      mResource.reset();
      mBundleRef.reset();
#endif

      mModule.reset();
      mAEffect = NULL;
   }
```

should be the responsibility of Wrapper, balancing what it does in load.  Make those lines a non-virtual function; call it in the destructor of LV2Wrapper, and also on the failure path of `Load()`.

And I think you defined virtual `callSetProgram` to avoid dependending on `mSlaves`.  But `Load()` only needs `callSetProgram` to initialize a plug-in or an instance which will not yet have any slaves.  So I think you don't need this function, and can instead let `Load` call `callSetProgramB`

pietro68 replied:
ok, done

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -563,4 +563,12 @@ class VSTEffectsModule final : public PluginProvider
+class VSTEffectValidator final : public DefaultEffectUIValidator
+{
+public:
+    VSTEffectValidator(EffectUIClientInterface& effect, EffectSettingsAccess& access);
```
Might simplify it with an inherited constructor:
`using DefaultEffectUIValidator::DefaultEffectUIValidator `

pietro68 replied:
ok, done

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -3609,6 +3609,13 @@ VSTEffectValidator::VSTEffectValidator(EffectUIClientInterface& effect,
 VSTEffectValidator::~VSTEffectValidator() = default;
 
 
+VSTEffectInstance::VSTEffectInstance(StatefulPerTrackEffect& effect)
```
The commit that stubs `ProcessInitialize` and others is incomplete.  My fault for giving an incomplete description for step 4.

`ProcessFinalize` should also be moved to the instance, and you will know all is correct when the effect also inherits `PerTrackEffect` not `StatefulPerTrackEffect`.

Please figure out step 4 as a whole and squash all the 4.x commits.

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -987,9 +987,7 @@ size_t VSTEffect::GetBlockSize() const
-   if (mUseLatency)
-      return mBufferDelay;
-   return 0;
+   return GetLatencyB();
```
`VSTEffectInstance` will also need to define all the other pure virtuals that were added to `VSTEffectWrapper`.

Some of the new functions should be empty stubs, others should have the lines from the corresponding VSTEffect functions moved into them, and data members should be moved too as needed to compile it.

Or a third possibility is that some members should be moved into VSTWrapper and the functions become non-virtual.

Which of the new functions of `VSTEffectInstance` should be empty?  I think NeedIdle, UpdateDisplay, SizeWindow all serve UI purposes only so should be empty.

What about the rest of them?  I'm not decided yet.

Paul-Licameli replied:
The virtual functions with non-void return are `GetTimeInfo` and `GetSampleRate`.  If we have a vst handle for the plug-in distinct from the instances, will it need to report meaningful values out of `AudioMaster` or will those cases never be invoked?

I'm not sure of the answer.  But I suggest this, and let experiment see if there is a problem:

Make default implementations of these non-pure virtuals to return nullptr at 44100.

Make overrides in the Instance class.

Move the member mTimeInfo to the instance.

Paul-Licameli replied:
It's unclear what `GetProcessLevel` is supposed to report to the plug-in, but the legacy code in fact never returned anything but 1 because mProcessLevel is never changed.

I say move mProcessLevel into the Wrapper class and provide a default implementation to the non-pure virtual.  Maybe we will figure out the need for an override later.

Paul-Licameli replied:
I think the `void`-returning `SetBufferDelay` and `Automate` are meaningful only for instances.  Make default implementations that do nothing.  Override them in the instance.

pietro68 replied:
Automate also references mMaster, so to have it a member of the instance, I had to move it to the wrapper.

I think I implemented all your suggestions here? please check commit "4.3 Implement some virtuals in the instance"

Paul-Licameli replied:
You did not remove `GetLatencyB` or undo the move of the two data members `mUseLatency`, `mBufferDelay`

Paul-Licameli replied:
If you examine how mMaster is used, it's really just a boolean that is true when an instance (a VSTEffect) is NOT the master.

Paul-Licameli replied:
You need mMaster in Automate, but the wrapper has a no-op override of Audomate, so there should be no need to put mMaster in the Wrapper.  You also moved mTimeInfo into the Wrapper but it should be in the Instane.

pietro68 replied:
> You need mMaster in Automate, but the wrapper has a no-op override of Automate, so there should be no need to put mMaster in the Wrapper.

but mMaster is set and used in 3 different VSTEffect methods - that's why I put it in the wrapper. I probably could rework the code in a way that VSTEffect does not need it anymore, but I'd leave that for later

pietro68 replied:
> You also moved mTimeInfo into the Wrapper but it should be in the Instance.

`mTimeInfo` is again written/used in different methods of `VSTEffect`. Should I figure out how to move these uses of it from `VSTEffect` to `VSTEffectInstance` as part of step 4?

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -199,6 +210,65 @@ struct VSTEffectWrapper : public VSTEffectLink, public XMLTagHandler
+   {
+      if (mUseLatency)
+         return mBufferDelay;
+      return 0;
```
Last comment for this round of review!

This definition of `GetLatencyB` seems weird to me.  I think you do it because the `SetBufferDelay` virtual function must be defined for all Wrappers.  But I say again, make it non-pure and let it do nothing in the base class.  Let the Instance override it properly.  Then, that leaves no need to move the two data members into the wrapper.  They belong in the Instance only.

pietro68 replied:
I commited part of your suggestion in 4.5. `mUseLatency` can not be moved to the instance, because it is referenced by `VSTEffect::ShowOptions` - any idea on how to solve that?

Paul-Licameli replied:
Just duplicate that boolean in the instance.  Make it a constructor argument.

pietro68 replied:
In commit 4.6 I moved `mUseLatency` and `mBufferDelay` in the instance, leaving a copy of `mUseLatency` in the Effect class which is passed to the instance at construction time.

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -101,6 +101,17 @@ struct VSTEffectSettings
+
+   // This constructor and SetPath are meant to be used when creating a VSTEffectInstance
+   VSTEffectWrapper() {}
+   void SetPath(const PluginPath& path)
```
Ah, one more comment.  Honestly the last.

I don't think this function is necessary.

pietro68 replied:
That's right, I could add the PluginPath as an arg to the Instance constructor. Done in commit 4.4

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -1822,6 +1821,28 @@ void VSTEffect::Unload()
       mAEffect = NULL;
    }
 
+   ResetModuleAndHandle();
```
This call within `Unload` is unnecessary.

pietro68 replied:
ok, removed

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -1822,6 +1821,28 @@ void VSTEffect::Unload()
+      mAEffect = NULL;
+   }
+
+   ResetModuleAndHandle();
```
This too is unnecessary.

pietro68 replied:
ok, removed

pietro68 commented:
Please note: "effect inherits PerTrackEffect not StatefulPerTrackEffect" is specified originally as part of step 4; but once I changed that inheritance, the code would not build until I implemented the most parts of step 5. - this is why it is part of step 5.

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -288,6 +288,38 @@ struct VSTEffectWrapper : public VSTEffectLink, public XMLTagHandler
 
    VSTEffect* mMaster;     // non-NULL if a slave
+
+#if defined(__WXMAC__)
```
Please do an interactive rebase, squashing this commit onto the second one, so that all commits are buildable.

pietro68 replied:
ok, now done and force-pushed

pietro68 replied:
oops I should delete the "build fix for OSx" commit - it's only whitespaces now

Paul-Licameli commented:
"removed unnecessary calls to ResetModuleAndHandle" might be squashed earlier

Paul-Licameli commented:
> Please note: "effect inherits PerTrackEffect not StatefulPerTrackEffect" is specified originally as part of step 4; but once I changed that inheritance, the code would not build until I implemented the most parts of step 5. - this is why it is part of step 5.

The problem, as I explained before, was that I omitted before part of what needed to be done in step 4.  That is, move `ProcessFinalize` also into the Instance.

Your step 5.1 simply deletes that function from the Effect but does not restore the equivalent into the Instance.  That may be losing an important side effect.

So I can't accept this pull request yet.

Paul-Licameli reviewed (changes requested):
This still isn't right and the fault is partly mine for not stating the steps in the issue clearly enough.  Please carefully re-read the updated steps 4 and 5.

For step 4: ProcessFinalize must also move into the instance.  The Effect stops inheriting StatefulPerTrackEffect but inherits TWO bases instead:  PerTrackEffect and StatefulEffectBase (which are also the direct bases of StatefulPerTrackEffect).  Just squash all the 4.x steps into one;  do not define ProcessBlock in the Wrapper because you don't need to shared it with the Effect.  Do not move callProcessReplacing into the Wrapper.

For step 5:  It is here that the Effect loses its other base class StatefulEffectBase, and other virtual functions of the effect for realtime move into the instance.

pietro68 commented:
I redid step 4. from scratch - hopefully it is not too complicated, I could not split it into subparts.

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -101,6 +101,15 @@ struct VSTEffectSettings
+      : mPath(path)
+   {}
+
+   ~VSTEffectWrapper()
```
I just noticed in review today, the omission of `virtual` from this destructor, though there are other virtual functions.

It might look nicer to add it.  It's not in fact important, so long as there are never yet any unique or shared pointers to a VSTEffectWrapper.

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -1341,7 +1355,8 @@ std::unique_ptr<EffectUIValidator> VSTEffect::PopulateUI(ShuttleGui &S,
    pParent->PushEventHandler(this);
-   return std::make_unique<DefaultEffectUIValidator>(*this, access, pParent);
+
+   return std::make_unique<VSTEffectValidator>(*this, access, pParent);
```
The minimal validator class doesn't yet appear to be necessary for the VST2 transformation, unlike the case for AudioUnits and VST3.  But no harm in keeping the commit that adds it.

pietro68 replied:
Well actually at this point I need it, because I pass the instance it belongs to in its constructor, to retrieve it later in ::ShowClientInterface

Paul-Licameli commented:
I wrote earlier "Also in step 4, the PowerOn() in BuildFancy() should be changed to call it on the Instance, not on this.
Pass the Instance into BuildFancy() from PopulateUI()."

The latest push doesn't do this any more.  Please restore that change.

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -199,6 +208,135 @@ struct VSTEffectWrapper : public VSTEffectLink, public XMLTagHandler
+   intptr_t mCurrentEffectID {};
+
+   bool mWantsIdle{ false };
+   std::unique_ptr<VSTEffectTimer> mTimer;
```
It is not necessary to move these two into the Wrapper and I think they will eventually be in the Instance.

pietro68 replied:
Correct. Now moved back in the effect.

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -212,15 +350,24 @@ using VSTEffectArray = std::vector < std::unique_ptr<VSTEffect> > ;
+// Temporary class allowing an intermediate phase in the transformation.
+// It is needed because the instance constructor needs something which is
+// both a PerTrackEffect and a StatefulEffectBase
+class TempClass : public PerTrackEffect
```
The existing `StatefulEffect` class might also serve the purpose

pietro68 replied:
Hmm... StatefulEffect does not inherit PerTrackEffect. I gave it a go anyway, and I got compiler errors. Never mind, this is only a temporary class that should disappear at point 5.

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -212,15 +350,24 @@ using VSTEffectArray = std::vector < std::unique_ptr<VSTEffect> > ;
 ///
 ///////////////////////////////////////////////////////////////////////////////
 class VSTEffect final
-   : public StatefulPerTrackEffect
```
The `#include` at top of the file can now be `PerTrackEffect.h`, not needing the stateful subclass

pietro68 replied:
ok, done

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -952,7 +950,8 @@ std::shared_ptr<EffectInstance> VSTEffect::DoMakeInstance()
    mBlockSize = mUserBlockSize;
-   return std::make_shared<Instance>(*this);
+ 
+   return std::make_shared<VSTEffectInstance>(*this, mPath);
```
An earlier version of this PR copied mUseLatency into the Instance in its constructor.  That change should remain.

What I should have said too:  the block size, and the user block size, should also be passed into the constructor of the instance.

Also the array of slaves used to be passed as a reference into the constructor.  Why no longer?  Do that too.

pietro68 replied:
mUseLatency, block size, user block size are now passed into the constructor of the instance and saved there.

I can not remember ever passing the array of slaves to the instance? and I wonder why I should do it now. I converted slaves to be instances. Why would a slave/instance need to know the other slaves/instances?

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -1038,7 +1037,14 @@ bool VSTEffect::ProcessFinalize()
+
+bool VSTEffectInstance::ProcessFinalize()
+{
+   return VSTEffectWrapper::DoProcessFinalize();
```
I don't like `DoProcessFinalize`, `DoProcessInitialize`, or `ProcessBlockB`.  Too much is stil being done in the wrapper class, and therefore the Effect still has more state that it should.  The purpose of this PR should be to remove as much effect state as possible, identifying the parts that should be per-instance state.

And I think the reason you thought you needed this is:  you let the slaves remain VSTEffect objects, and did not change the type of the slaves to be VSTEffectInstance.  But you can do this.

The array of slaves remains temporarily in the Effect, until step 5 is done.  But the virtual functions needed in the slaves are only `ProcessFinalize` and `ProcessBlock`, now in the Instance class, and other functions already in the Wrapper.

pietro68 replied:
I now implemented the "middle class" VstInstanceBase as we discussed, please have a look there.

Paul-Licameli replied:
I do not yet see a force-push with that update

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -199,6 +208,141 @@ struct VSTEffectWrapper : public VSTEffectLink, public XMLTagHandler
+   virtual void UpdateDisplay() = 0;
+   virtual void SizeWindow(int w, int h) = 0;
+
+   virtual VstTimeInfo* GetTimeInfo()
```
I would prefer to have these virtual functions defined non-inline, with comments inside the class about what they do (which is mostly nothing)

pietro68 replied:
I made them inlinenow.

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -264,19 +417,17 @@ class VSTEffect final
-      const float *const *inBlock, float *const *outBlock, size_t blockLen)
-      override;
+
+   bool Process(EffectInstance& instance, EffectSettings& settings)
```
This function should be marked `override` and its definition should be non-inline

pietro68 replied:
ok, done

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -952,7 +941,8 @@ std::shared_ptr<EffectInstance> VSTEffect::DoMakeInstance()
    GetConfig(*this, PluginSettings::Shared, wxT("Options"),
```
If all is done right, `DoMakeInstance` and the `const_cast` should no longer be needed.
`mUseLatency` and `mUserBlockSize` don't need to be member variables of the effect.  They can become local variables of this function.  But then they must be passed to the constructo of Instance which stores them.

But this may need to wait until step 5.

pietro68 replied:
I now pass them to the Instance constructor. As you said, later in step 5. I could remove them from the effect.

pietro68 commented:
> I wrote earlier "Also in step 4, the PowerOn() in BuildFancy() should be changed to call it on the Instance, not on this. Pass the Instance into BuildFancy() from PopulateUI()."
> 
> The latest push doesn't do this any more. Please restore that change.

ok, restored

pietro68 commented:
All of what we discussed yesterday is now in the last push I did, commit 4.

Paul-Licameli commented:
The latest push looks good, but it awaits completion with step 5.  A few comments:

- `VSTEffectValidator::mIntance` might be private.
- I was surprised to see `AudioMaster` moved into `VSTInstanceBase`.  I thought at first, that's wrong, it should remain in the Wrapper, and the Validator may need to use it too.  But your comments suggest, we will split the `AudioMaster` function into multiple functions, and pass different functions to `pluginMain`.  I now think that's a good alternative solution too.
- `ProcessBlockB ` doesn't need to be separated from `ProcessBlock ` as a member function of `VSTInstanceBase`, now that slaves are also Instances.  (But not so, yet, for the initialize and finalize functions.)
- Where is `VSTEffectInstance::mNumChannels` used?  I can't find it.  You might just delete that, and the functions that get and set it.
- `callSetChunk` (which has two overloads) is no longer called anywhere!  (And the four-argument overload in fact became unused already at commit 2df527f6d72c92c5cf5c8cc679ac6122a03c4649). So, just delete those, and rename the `callSetChunkB` functions as `callSetChunk`.
- I think `VSTInstanceBase::Automate` is wrong... I think the `VSTInstance` needs to override it, and call `callSetParameter` (not B), and so, the Instance needs, transitionally, a reference to the array of slaves.
- Why is `mTimeInfo` still in the Effect class?  Only the Instance and the slaves, also Instances, need it.  Once you address the point about `ProcessBlockB`, this will become clearer.  Remove it from the Effect.

Please do an interactive rebase addressing all of these, force push it, and then proceed to step 5.

pietro68 commented:
> * Why is `mTimeInfo` still in the Effect class?  Only the Instance and the slaves, also Instances, need it.  Once you address the point about `ProcessBlockB`, this will become clearer.  Remove it from the Effect.

`mTimeInfo` is actually in the InstanceBase class, not the Effect class; it is referenced by:
- `GetTimeInfo()`  called from the vst callback
- `GetSampleRate()` called from the vst callback
- `VSTEffect::VSTEffect` 

As long as the vst callback is called for the effect too, I can not move `mTimeInfo` in the Instance only. I actually wonder if I can move it there even after I split the vst callback in two parts (one for the effect, one for the instance) because maybe the callback for the effect will still need to call `GetTimeInfo` and/or `GetSampleRate`? this because the vst GUI might want to draw something that depends on the samplerate (think of a band equalizer for instance)

Paul-Licameli commented:
> > * Why is `mTimeInfo` still in the Effect class?  Only the Instance and the slaves, also Instances, need it.  Once you address the point about `ProcessBlockB`, this will become clearer.  Remove it from the Effect.
> 
> `mTimeInfo` is actually in the InstanceBase class, not the Effect class; it is referenced by:
> 
> * `GetTimeInfo()`  called from the vst callback
> * `GetSampleRate()` called from the vst callback
> * `VSTEffect::VSTEffect`
> 
> As long as the vst callback is called for the effect too, I can not move `mTimeInfo` in the Instance only. I actually wonder if I can move it there even after I split the vst callback in two parts (one for the effect, one for the instance) because maybe the callback for the effect will still need to call `GetTimeInfo` and/or `GetSampleRate`? this because the vst GUI might want to draw something that depends on the samplerate (think of a band equalizer for instance)

I thought we decided it is not needed in the Effect. If for GetTimeInfo and GetSampleRate, the effect can simply inherit the trivial implementations, then mTimeInfo can belong to Instance only.

Paul-Licameli commented:
> > > * Why is `mTimeInfo` still in the Effect class?  Only the Instance and the slaves, also Instances, need it.  Once you address the point about `ProcessBlockB`, this will become clearer.  Remove it from the Effect.
> > 
> > 
> > `mTimeInfo` is actually in the InstanceBase class, not the Effect class; it is referenced by:
> > 
> > * `GetTimeInfo()`  called from the vst callback
> > * `GetSampleRate()` called from the vst callback
> > * `VSTEffect::VSTEffect`
> > 
> > As long as the vst callback is called for the effect too, I can not move `mTimeInfo` in the Instance only. I actually wonder if I can move it there even after I split the vst callback in two parts (one for the effect, one for the instance) because maybe the callback for the effect will still need to call `GetTimeInfo` and/or `GetSampleRate`? this because the vst GUI might want to draw something that depends on the samplerate (think of a band equalizer for instance)
> 
> I thought we decided it is not needed in the Effect. If for GetTimeInfo and GetSampleRate, the effect can simply inherit the trivial implementations, then mTimeInfo can belong to Instance only.

If experiment shows that those cases in AudioMaster are not reached for the Effect object, but only for the Instance, then
the simplification of moving mTimeInfo out of Effect could be done and would be desirable.

But we might leave that until we do such experiment.

pietro68 commented:
> * I think `VSTInstanceBase::Automate` is wrong... I think the `VSTInstance` needs to override it, and call `callSetParameter` (not B), and so, the Instance needs, transitionally, a reference to the array of slaves.

I'd like to think out loud a bit now. Automate is called from the vst callback, sometimes on the Effect, sometimes on the Instance. I put some breakpoints and I observed this:

1) when instantiating an effect 
--> Automate is called on the effect, nothing happens because mMaster is null and mSlaves is empty

2) when moving a slider on the effect gui --> as above

3) when pressing play 
--> Automate is called on the instance

4) when clicking "apply" (i.e. destructive processing) 
--> Automate is called first on the effect (mMaster=null mSlaves={}) and then on the instance

5) when moving a slider while playing 
--> Automate is called on the effect, mMaster=null but this time mSlaves has one item

My conclusions (but I might be wrong): 

a) we need both overrides of Automate (for Effect and Instance), and when I will split the callback, both of them will need to call Automate when requested to. 

b) mMaster can be removed (it is always nullptr in my tries above)

c) I could do as you say, i.e. pass the array of slaves to the instance and have Instance::Automate call callSetParameter (not B) on them, but, is it really needed? without this, everything still works and it is a temporary thing anyway

pietro68 commented:
> If experiment shows that those cases in AudioMaster are not reached for the Effect object, but only for the Instance, then the simplification of moving mTimeInfo out of Effect could be done and would be desirable.
> 
> But we might leave that until we do such experiment.

Did a little experiment: using a simple plugin like gain, `GetTimeInfo()` is never reached, neither for the effect nor for the instance. But already using an equalizer plugin, at least `GetTimeInfo()` is reached, and for the effect only. 
So I would say, at least `GetTimeInfo()` should remain for the effect, and if we want to play it safe, also` GetSampleRate()`.

pietro68 commented:
A note about my commits for step 5. :

I said earlier (and wrote in code comments) that I would split the vst callback function in two parts, one for the instance and one for the effect.

Since then I changed my mind and opted for a more conservative solution, which is: keep the vst callback in the wrapper and have it call virtual methods. I think doing so is cleaner and poses less risks of breaking things.

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -3599,4 +3597,42 @@ EffectSettings VSTEffect::MakeSettings() const
+}
+
+int VSTEffectWrapper::GetProcessLevel()
+{
```
This is the one virtual function of `VSTEffectWrapper` that we can define in the base and then not override again, or so we think now.

So the first commit might also remove `VSTEffect::GetProcessLevel()`

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -3607,6 +3608,18 @@ EffectSettings VSTEffect::MakeSettings() const
+   VSTEffectInstance&       instance,
+   EffectUIClientInterface& effect,
+   EffectSettingsAccess&    access,
+   wxWindow*                pParent
```
The commit that defines the Instance adds a reference to Instance in the Validator.  It also adds three members that are not needed.  They only duplicate protected members of the base class.

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -287,29 +270,103 @@ struct VSTEffectWrapper : public VSTEffectLink, public XMLTagHandler
+   // be able to move the callback and the method it calls into the instance class.
+   //
+   // Actually: the callback will have to be split in two - one part as the callback
+   // for the instances, and the remaining part as the callback for the effect
```
Is this part of the comment still correct?

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -1087,11 +1086,11 @@ bool VSTEffect::RealtimeAddProcessor(
       callDispatcher(effBeginSetProgram, 0, 0, NULL, 0.0);
       for (int i = 0; i < mAEffect->numParams; i++)
-         slave->callSetParameter(i, callGetParameter(i));
+         slave->callSetParameterB(i, callGetParameter(i));
```
I did not mention this before, but if you look for the analogous uses of `StoreSettings` on AudioUnit instances, one of them happens here.  In step 4 you might make a similar change here too, eliminating the loop over parameters, and the beginnin and end of set-program, which would happen inside `StoreSettings`.

pietro68 replied:
OK done that in commit 4.2 (yet to be pushed at this moment) - you will see that to make it work I also had to add a call to `FetchSettings` inside `Automate`.

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -3620,47 +3627,74 @@ VSTEffectValidator::VSTEffectValidator
 {
-   //return mProcessLevel;
-   return 1;
+   callSetParameterB(index, value);
```
Automate() is changed to use `callSetParameterB`, losing the loop over slaves.  That seems like a wrong change in side-effects.

You should call `callSetParameter`.   But that doesn't exist in `VSTInstanceBase`.

The fix should then be:  move Automate() to the Instance class in step 4, which is what the header file says should happen.

Also move `callSetParameter` (which includes the loop) into Instance, out of the Effect object.

But then `VSTEffect::OnSlider` won't compile.  But the solution for that will be, the Validator object will call `callSetParameter` on the Instance.  That function will ultimately be in the validator.

So make this temporary fix here in step 4:  Let the Effect retain a pointer to the last constructed Validator.  Assume there is never more than one Valdator at once and that its lifetime will include any call to `OnSlider`, which will happen only while a non-fancy, plain UI dialog is open.

pietro68 replied:
I was wrong when I said that Automate has to belong to the instance. 
It must indeed belong to the Effect, because changes in sliders are intercepted by the Effect vst callback.

That being said, yes it is still wrong that Automate() loses the loop over slaves. In commit 4.3 I fix that by having Automate use callSetParameter and that required that mInstances move to the instancebase class.

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -2204,7 +2221,7 @@ void VSTEffect::callSetProgram(int index)
    callDispatcher(effSetProgram, 0, index, NULL, 0.0);
    for (const auto &slave : mSlaves)
-      slave->callSetProgram(index);
+      slave->callSetProgramB(index);
```
Can the distinction between `callSetProgram` and `callSetProgramB` now be eliminated?  I think so, but it needs some longer explanation from me.

pietro68 replied:
The short answer is: not, when doing realtime processing, at least at this stage (I mean, step 4)

Why? I tried the following: 
1) place breakpoint in CallSetProgram
2) choose a factory preset: when we reach CallSetProgram, mSlaves is empty

So if we did only destructive processing, then we could go with callSetProgramB only (i.e. the version which ignores slaves)

but:
if I start realtime processing and then choose a factory preset, when I reach callSetProgram, mSlaves is not empty - it has one item and I guess it is the item doing the realtime processing.

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -1037,7 +1039,14 @@ return GuardedCall<bool>([&]{
+
+bool VSTEffectInstance::ProcessFinalize() noexcept
+{
+   return VSTInstanceBase::DoProcessFinalize();
```
I think there is no longer any need to make `DoProcessFinalize` a separate function.

pietro68 replied:
Removing it at this stage is not so simple, because DoProcessFinalize is in VSTInstanceBase and it is used by VSTEffect; but VSTEffect does not know VSTEffectInstance... we'd need to wait until step 5

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -1128,21 +1128,31 @@ bool VSTEffect::RealtimeResume()
+      // If we assume that the user might be moving knobs during realtime processing and wants
+      // to hear how the sound changes, we must do this then
+
+      auto guard = std::lock_guard{ GetEffect().mSettingsMutex };
```
Do not do this.

We don't want to do mutex locking and unlocking in a worker thread.

We will trust `settings` to be a copy of settings, properly updated in an atomic fashion, by the `EffectSettingsAccess` object.

Yes, strictly speaking you are right that a data race on the `VSTEffectSettings` stored in the effect, is possible while the transition to statelessness is not complete, and you might remove this mutex locking then.

But I think that is excessive caution.  I didn't transform AudioUnit effects with similar caution and I don't recommend you do it for VST.

pietro68 replied:
As I told you during our talk, without introducing this mutex, I always ran into a crash when I moved a slider while processing realtime. This is because mSettings was written in the main thread (when moving the slider), while it was read in the worker thread. 
But I will add a comment about the mutex being a temporary thing until the validator is fully implemented.

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -2191,18 +2188,14 @@ float VSTEffectWrapper::callGetParameter(int index) const
-   {
-      for (const auto& slave : mSlaves)
-         slave->callSetParameterB(index, value);
-   }
```
Your step 5.1 removes these lines.  I already explained how it should be done in step 4, eliminating the need for two versions of the set-parameter function.

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -3650,20 +3641,31 @@ void VSTInstanceBase::SizeWindow(int w, int h)
+      // thread, but the storing of the settings passed to ::RealtimeProcess
+      // happens in the worker thread
+      auto guard = std::lock_guard{ mSettingsMutex };
+      FetchSettings(mSettings);
```
You made these changes in 5.1.  I don't like the mutex.

Does experiment show that this is called on the main thread only in response to slider changes?

I need the answer to that, before I can recommend the correct thing  to do.

pietro68 replied:
Please see my comment above https://github.com/audacity/audacity/pull/3278#discussion_r946051875

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -605,6 +582,27 @@ class VSTEffectInstance final : public StatefulEffectBase::Instance,
+   size_t GetBlockSize() const override;
+
+   // this not?
+   //bool Process(EffectInstance& instance, EffectSettings& settings) override;
```
There is nothing in the instance now corresponding to `Process`.  Instead you trust `PerTrackEffect` to do destructive effect processing in its `Process`, calling the `ProcessBlock` function of the instance.  And after other recent refactorings, that happens in member functions of an `EffectStage` object.

pietro68 replied:
ok

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -785,9 +785,8 @@ void VSTEffect::ResourceHandle::reset()
-:  VSTInstanceBase(path),
-   mMaster(master)
+VSTEffect::VSTEffect(const PluginPath & path)
+:  VSTInstanceBase(path)
```
I think step 5.3 makes more sense squashed with step 4

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -1182,7 +1177,7 @@ int VSTEffect::ShowClientInterface(
       // normal or realtime processing begins
       mBlockSize = 8192;
-      DoProcessInitialize(mProjectRate);
+      vstValidator->GetInstance().DoProcessInitialize(mProjectRate);
```
This change might be squashed onto step 4, making 4 more correct

pietro68 replied:
Ok, done in commit 4.5

Paul-Licameli commented:
> I was wrong when I said that Automate has to belong to the instance. It must indeed belong to the Effect, because changes in sliders are intercepted by the Effect vst callback.
> 
> That being said, yes it is still wrong that Automate() loses the loop over slaves. In commit 4.3 I fix that by having Automate use callSetParameter and that required that mInstances move to the instancebase class.

I'm still not satisfied.  As I explain:  Both in Automate(), and in OnSlider() (which will both ultimately go to the Validator class), callSetParameter should be invoked, but NOT in `this`, but instead on an Instance object.

That is the Instance stored in the minimal Validator class now; and, until the Validator is complete, the Effect will keep a pointer to the active Validator.

pietro68 commented:
> I'm still not satisfied. As I explain: Both in Automate(), and in OnSlider() (which will both ultimately go to the Validator class), callSetParameter should be invoked, but NOT in `this`, but instead on an Instance object.
> 
> That is the Instance stored in the minimal Validator class now; and, until the Validator is complete, the Effect will keep a pointer to the active Validator.

I can not do this because:

1) `::Automate` comes in two versions:

- `VSTInstanceBase::Automate`  which has no access to members in the Effect
- `VSTEffect::Automate` it is actually calling callsetparameterB


2) I tried doing it in `::OnSlider`, but then it breaks this functionality:  as I move sliders of the plain GUI while realtime processing, I do not hear changes in the sound anymore.

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -3594,6 +3595,18 @@ EffectSettings VSTEffect::MakeSettings() const
+   VSTEffectInstance&       instance,
+   EffectUIClientInterface& effect,
+   EffectSettingsAccess&    access,
+   wxWindow*                pParent
```
I suggest it again:  at step 2, you don't need `effect`, `access`, `pParent` members because you can just use the protected members of the base classes of `VSTEffectValidator.`

Paul-Licameli commented:
I am reviewing the latest force-push and so far I am satisfied with the 4.x steps.  They can be squashed into one in your next force-push.

Paul-Licameli commented on src/effects/VST/VSTEffect.h:
```diff
@@ -150,6 +151,8 @@ struct VSTEffectWrapper : public VSTEffectLink, public XMLTagHandler
 
    VSTEffectSettings mSettings;  // temporary, until the effect is really stateless
+   std::mutex mSettingsMutex;    // to avoid read/write races on mSettings - this is needed temporarily
+                                 // and will be removed when the Validato will be implemented
```
"Validator"

pietro68 replied:
ok

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -1154,6 +1165,7 @@ int VSTEffect::ShowClientInterface(
    //   mProcessLevel = 1;      // in GUI thread
 
    VSTEffectValidator* vstValidator = static_cast<VSTEffectValidator*>(validator);
+   mValidator = vstValidator;
```
Step 5.1 adds this assignment, but never uses the member, and never reassigns it back to null.  But I'm still progressing through the 5.x commits, which may be squashed into one if I approve them.

pietro68 replied:
Correct. I introduced `mValidator` as a way to call `callSetParameter` on the Validator's instance, in `::OnSlider`, as you previously suggested (or so I understood). Then I found out that things work also without that call (and I left the code there commented out) so I could delete `mValidator` now. Shall I do it?

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -374,7 +374,7 @@ unsigned VSTEffectsModule::DiscoverPluginsAtPath(
          //to initialize EffectDefinitionInterface part, which includes
          //properly formatted plugin path
-         VSTEffect subeffect(wxString::Format("%s;%d", path, id), &effect);
+         VSTEffect subeffect(wxString::Format("%s;%d", path, id));
```
In step 5.3, add:
`subeffect.Load()`
to preserve the behavior.

pietro68 replied:
Good catch! ok

Paul-Licameli commented on src/effects/VST/VSTEffect.cpp:
```diff
@@ -2013,7 +2009,7 @@ int VSTInstanceBase::GetProcessLevel()
 }
 
-void VSTInstanceBase::PowerOn()
+void VSTEffectWrapper::PowerOn()
```
In step 5.4, think `PowerOn` and `PowerOff` should not move to Wrapper.  They are only for the Instance.

Power comes ON, only in an instance.  Note `vstEffInstance.PowerOn()` -- it's not done on This.

But there are calls to `PowerOff` done on the Effect.

The right thing is to change those.

Where `CloseUI` does `PowerOff`, do it instead to the Instance stored in the Validator.  Which is why `VSTEffect::mValidator` finds a use after all.

Als in `CloseUI`, reassign `mValidator` to null when done.

The `PowerOff()` in the unloading of the effect should just be deleted.

Paul-Licameli commented:
Please rebase onto latest master, squash steps 4 into one, and squash steps 5 into another commit.  Address in step 5 the two reqeusts for change.  Then, this will be approved, and you will move on to the Validator.

Paul-Licameli commented:
Less important, `mUseLatency` might also move from Wrapper to Instance

Paul-Licameli reviewed (approved):
Steps 4.x and 5.x might be squashed into just two commits before merging.  All else looks good now.
